### ISSUE TITLE
Gradient propagation through channel

### ISSUE THREAD

**Issue by Rassibassi** [Fritschek/Wireless_encoding_with_MI_estimation#1]
Hi,

cool paper and idea :) 

Some months ago, I tried to implement your method, but could not make it work myself. So thanks for the example implementation.

However, I have a question regarding the following statement in the paper:

> We use this estimator to optimize the encoder for a maximized mutual information, only relying on channel samples.

In the example code, to the best of my knowledge, gradient propagation through the channel is allowed. So not only relying on channel samples, but also on the channel model. When I tried to explicitly dissallow such gradient propagation, it seems to fail.

I only changed one line:
```python
channel_layer = keras.layers.Lambda(lambda x: 
                    tf.stop_gradient(x + tf.random.normal(tf.shape(x), mean=0.0, stddev=noise_std)))
```

see here: https://github.com/Rassibassi/Wireless_encoding_with_MI_estimation/blob/master/TF2_MINE_Enc_em.ipynb

If you have time, could you please elaborate? Is there something I don't understand?

Thanks,
Rasmus

**Comment by Fritschek** [comment]
Hi Rasmus,

Thanks for your interest in our paper.
That's a really good question :).
The code was just a matter of convenience. Currently, Im thinking about how
to implement this better.
I think stop gradients is doing the wrong thing here, since it makes the
channel a constant, and this results in something different.
Since, we still need the channel dynamics (however, we just need samples at
"run time").
The encoder needs to be trained, such that it sees that I(x,y) is changing
in a certain way, with its encoding weights.
Better way would be to create a look-up table of values on the go, with TF
not seeing the generating function, which are then put into MI.
That should work, I guess.
Need to think about a good way to implement this though...

Best regards,
Rick

**Comment by Rassibassi** [comment]
Hi again,

thank you for your answer. Yet, I am not convinced :)

`tf.stop_gradient` does preserve the channel dynamics, e.g. i.i.d. noise samples are added within the channel for every sample. The output of the channel is not constant, but becomes as if declared as `tf.constant` for every sample.

Another way of restricting TF's automatic differentiation algorithm to propagate a gradient through the channel, is to simply apply the channel calculations in numpy. This simulates a physical channel to obtain the channel samples, and at no time the channel is part of TF. This is similar to your comment "TF not seeing the generating function".

However, with the channel implemented in numpy, the example code still fails :( See here for the numpy channel example:
https://github.com/Rassibassi/Wireless_encoding_with_MI_estimation/blob/master/TF2_MINE_Enc_em-numpyChannel.ipynb

Any ideas?

Best,
Rasmus

**Comment by Fritschek** [comment]
Hi Rasmus,

I still think that your code is not doing what we need. As I understand it, stop_gradients is stopping the gradients from backpropagating, but we actually also need the gradients of the encoder, because this is what we train for. 
With the numpy channel its the same, even if we don't need the channel gradients itself, the chain rule is now interrupted before the channel.

So we need the channel function in the forward pass and a bypass in the backward pass.
This can be implemented with tf.grad_pass_through.
Taking your code example with the numpy channel, one can modify the y_recv line to
`                y_recv = tf.grad_pass_through(channel)(x_enc)
`
in the train_encoder function, in the encoder training sub-loop. This restores the learning capabilities, although one might need a bit of hyperparameter tuning to get the MI Pe curve really close to 16QAM for 12+ EbN0 db again.
Note that we don't need that pass through in the MI loops, MI train function or decoder train function, as these derivatives stop before the channel.

Anyways, I'm currently on vacation, so I dont have much time to look into this.
But, afterwards, I see if I can maybe implement this in a better way.
And will do some further testing and improve the code.
Thanks for pointing this out to me.

Best regards,
Rick

**Comment by Rassibassi** [comment]
Hi again,

implemented the `grad_pass_through` idea:
https://github.com/Rassibassi/Wireless_encoding_with_MI_estimation/blob/master/TF2_MINE_Enc_em-channel_passthrough.ipynb

It works now, however, the passthrough assumes the identity as gradient. Obviously, this works well for the AWGN channel. Not sure about other channels.

Hope you are safe and sound during these times :)

Best,
Rasmus

**Comment by Fritschek** [comment]
Hi,

I'm fine, Germany is now in partial lockdown. Unusual times.
I hope you are also well.

Totally forgot to update the repo.
I have now included the pass through in the encoder, and also added a Rayleigh channel comparison.
As I said before, pass through is not needed for MI, as the gradient is calculated for the MI variables, and therefore stops before the channel. So it doesn't matter/change the result.
Moreover, it's not obvious that it works well for the Gaussian channel, as the gradient is different if you just have the identity / leave out the channel.
Anyways, its also different for the Rayleigh channel, which is why I've added this comparison too.
As we claimed in our paper, it looks like that the MI information is enough to optimize the channel encoder.

I also added a few other fixes and improvements.

Best regards,
Rick
